overlord/devicestate/devicestate.go

```diff
@@ -1411,35 +1442,289 @@ func createRecoverySystemTasks(st *state.State, label string, snapSetupTasks []s
+	// places that IDs are used
+	if err := checkForSnapIDs(model, opts.LocalSnaps); err != nil {
+		return nil, err
+	}
+
+	// check that all snaps from the model are valid in the validation sets
+	if err := checkModelSnapsCanBePresentInValidationSets(model, valsets); err != nil {
+		return nil, err
+	}
+
+	// snaps required by validation sets must also be required by the model
+	if err := checkForRequiredSnapsNotRequiredInModel(model, valsets); err != nil {
+		return nil, err
+	}
+
+	tracker := snap.NewSelfContainedSetPrereqTracker()
+	offline := len(opts.LocalSnaps) > 0
+
+	var downloadTSS []*state.TaskSet
+	for _, sn := range model.AllSnaps() {
+		rev := revisions[sn.Name]
+
+		needsInstall, err := snapNeedsInstall(st, sn.Name, rev)
+		if err != nil {
+			return nil, err
+		}
+
+		if !needsInstall {
+			info, err := snapstate.CurrentInfo(st, sn.Name)
+			if err != nil {
+				return nil, err
+			}
+			tracker.Add(info)
+			continue
+		}
+
+		if offline {
+			info, err := offlineSnapInfo(sn, rev, opts)
+			if err != nil {
+				return nil, err
+			}
+			tracker.Add(info)
+			continue
+		}
+
+		const userID = 0
+		// TODO: this respects the passed in validation sets, but does not
+		// currently respect refresh-control style of constraining snap
+		// revisions.
+		ts, info, err := snapstateDownload(context.TODO(), st, sn.Name, dirs.SnapBlobDir, &snapstate.RevisionOptions{
+			Channel:        sn.DefaultChannel,
+			Revision:       rev,
+			ValidationSets: valsets.Keys(),
+		}, userID, snapstate.Flags{}, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		tracker.Add(info)
+		downloadTSS = append(downloadTSS, ts)
+	}
+
+	warnings, errs := tracker.Check()
+	for _, w := range warnings {
+		logger.Noticef("create recovery system prerequisites warning: %v", w)
+	}
+
+	// TODO: use function from other branch
+	if len(errs) > 0 {
+		var builder strings.Builder
+		builder.WriteString("cannot create recovery system from model that is not self-contained:")
+
+		for _, err := range errs {
+			builder.WriteString("\n  - ")
+			builder.WriteString(err.Error())
+		}
+
+		return nil, errors.New(builder.String())
+	}
+
+	var snapsupTaskIDs []string
+	if len(downloadTSS) > 0 {
+		snapsupTaskIDs, err = extractSnapSetupTaskIDs(downloadTSS)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	chg = st.NewChange("create-recovery-system", fmt.Sprintf("Create new recovery system with label %q", label))
+	createTS, err := createRecoverySystemTasks(st, label, snapsupTaskIDs, opts)
 	if err != nil {
 		return nil, err
 	}
-	chg.AddAll(ts)
+
+	chg.AddAll(createTS)
+
+	for _, ts := range downloadTSS {
+		createTS.WaitAll(ts)
+		chg.AddAll(ts)
+	}
+
 	return chg, nil
 }
 
+func checkForSnapIDs(model *asserts.Model, localSnaps []LocalSnap) error {
+	for _, sn := range model.AllSnaps() {
+		if sn.ID() == "" {
+			return fmt.Errorf("cannot create recovery system from model with snap that has no id: %q", sn.Name)
```

Comment:
s/no id/no snap id/  , we tend to always prefix id with snap even if it's mostly clear, as we have many ids